Regent-FFT update

arjunkunna opened the pull request:
Update of the Regent-FFT API. The following features are supported:

- Supports CPU and single-GPU transforms for 1D, 2D, and 3D.
- Supports Real-to-Complex and Complex-to-Complex transforms for both CPU (complex64 only) and GPU (complex32 and complex64).
- Supports batched transforms: both R2C and C2C (complex32 and complex64) for GPUs, and R2C and C2C (only complex64) for CPUs.

Please refer to README for more information on usage. 

@elliottslaughter - I sent you an email as well. Do let me know if you have any comments on the code!

Commit:
transferring code from previous repo

Commit:
testing GHA

Commit:
change main.yml to master.yml

Commit:
change main.yml to master.yml

Commit:
Comments in execute_plan

Commit:
comments

Commit:
trigger GHA

arjunkunna committed:
Update README.md

Commit:
trigger GHA

Commit:
Merge branch 'master' of https://github.com/arjunkunna/regent-fft

Commit:
trigger GHA

Commit:
Remove loggingwrapper

Commit:
remove jemalloc

Commit:
remove logging wrapper source code

Commit:
r2c doesn't work

Commit:
update comments

Commit:
test.sh

arjunkunna committed:
Update README.md

Commit:
commit before PR

arjunkunna committed:
Update README.md

arjunkunna committed:
Update README.md

arjunkunna committed:
Update README.md

elliottslaughter commented on .github/workflows/master.yml:
```diff
@@ -0,0 +1,35 @@
+on: 
+  push:
+    branches: [master]
+    paths-ignore: ['**/*.md']
```
I think we can remove these lines? We want this to run on all branches.

elliottslaughter commented on .github/workflows/master.yml:
```diff
@@ -0,0 +1,35 @@
+    paths-ignore: ['**/*.md']
+  pull_request:
+    branches: [master]
+    paths-ignore: ['**/*.md']
```
Same.

arjunkunna replied:
Removed these

elliottslaughter commented on .github/workflows/master.yml:
```diff
@@ -0,0 +1,35 @@
+  pull_request:
+    branches: [master]
+    paths-ignore: ['**/*.md']
+  workflow_dispatch:
```
Not sure we need this?

arjunkunna replied:
Removed these too

elliottslaughter commented on .github/workflows/master.yml:
```diff
@@ -0,0 +1,35 @@
+    strategy:
+      matrix:
+        os: ['ubuntu-22.04'] # 'macos-11' # FIXME: macOS is broken: https://github.com/StanfordLegion/legion/issues/213
+        # debug: ['1', '0']
```
Clean up comments here. It's ok if we only run this on Linux.

arjunkunna replied:
Cleaned this up - removed comments

elliottslaughter commented on .github/workflows/master.yml:
```diff
@@ -0,0 +1,35 @@
+      matrix:
+        os: ['ubuntu-22.04'] # 'macos-11' # FIXME: macOS is broken: https://github.com/StanfordLegion/legion/issues/213
+        # debug: ['1', '0']
+        debug: ['0']
```
But we almost certainly should test this in debug mode?

arjunkunna replied:
Testing in debug now.

elliottslaughter commented on .github/workflows/master.yml:
```diff
@@ -0,0 +1,35 @@
+        shell: bash
+        env:
+          DEBUG: ${{ matrix.debug }}
+          USE_REGENT: '1'
```
Can you review these variables to make sure they are all relevant? E.g., I don't think we have a non-Regent mode in this library.

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,237 @@
+# Regent-FFT
 
-## Installation
+[![Regent-FFT Tests](https://github.com/arjunkunna/regent-fft-arjun/actions/workflows/main.yml/badge.svg)](https://github.com/arjunkunna/regent-fft/actions/workflows/master.yml)
```
Please update the link here.

arjunkunna replied:
Updated

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,237 @@
+
+Both `complex64` and `complex32` types are supported in GPU mode. CPU mode is
+only able to support `complex64`. It is possible to use `complex32` in CPU mode
+but it requires some additional setup - please contact me if that is of
```
Since this is merging into an upstream repo, let's say something like "please submit an issue if that is of interest".

arjunkunna replied:
Edited this.

elliottslaughter commented on README.md:
````diff
@@ -1,189 +1,237 @@
+Then, clone the repo:
+
+```shell
+https://github.com/arjunkunna/regent-fft.git
````
Update URL.

elliottslaughter replied:
Also, might as well write this out:

```
git clone ...
```

arjunkunna replied:
Edited

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,237 @@
-specific number of dimensions and data type you are interested
-in. Currently only 1 to 3 dimensions and `complex64` are
-supported. For example, to initialize with 1 dimension:
+### Executing program
```
Title Case?

arjunkunna replied:
Updated title case throughout Readme

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,237 @@
+as well.
 
-To initialize in non-distributed mode:
+#### 1. Link the `fft.rg` file and generate an interface
```
Instead of "link" I would say "import"

elliottslaughter replied:
Also, I see this is not in Title Case? For consistency should we match the rest of this file?

arjunkunna replied:
Edited and made title case

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,237 @@
-To initialize in non-distributed mode:
+#### 1. Link the `fft.rg` file and generate an interface
 
+- The first argument is the dimension - `int1d`, `int2d`, or `int3d`.
```
Minor nit but I'd use a colon `:` instead of a dash here.

arjunkunna replied:
Edited

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,237 @@
 
+- The first argument is the dimension - `int1d`, `int2d`, or `int3d`.
+- The second argument is the data type of the input - `complex64`, `complex32`,
+  `real`, or `double`.
```
Do you mean `float` instead of `real`?

arjunkunna replied:
Yep, thanks - edited this

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,237 @@
+- The first argument is the dimension - `int1d`, `int2d`, or `int3d`.
+- The second argument is the data type of the input - `complex64`, `complex32`,
+  `real`, or `double`.
+- The third argument is the data type of the input - `complex64` or `complex32`.
```
Output? I assume it is not input again.

arjunkunna replied:
Thanks for the catch - edited as well

elliottslaughter commented on README.md:
````diff
@@ -1,189 +1,237 @@
 fft1d.make_plan(r, s, p)
 ```
 
-**Important:** `make_plan` overwites the input and output regions `r`
-and `s`. This is mandated by FFTW, which Regent FFT uses on CPUs. In
-order to avoid overwriting data, the user must either initialize the
-plan prior to loading the regions with data, or else must create a
-temporary region (of the same size and layout as the real one) for use
-in initialization.
````
It looks like this note got deleted, but as far as I know it is still true. So we should probably be careful to keep it.

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,237 @@
+`ispace(<type>, N)`, where N is the size of the array, and `<type>` is either
+int1d/int2d/int3d depending on the dimension of the transform. The fieldspace of
+the region is the type supported by the transform - e.g, in a real-to-complex
+transform with doubles, the input array will have fieldspace `double` and output
+array will have fieldspace `complex64`.
```
Field polymorphism allows you to pass fields in larger field spaces as well.

arjunkunna replied:
Added a comment to note this

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,237 @@
-`__demand(__inline)` task that internally performs an index launch
-over the machine to initialize `p`.
+`make_plan` is a `__demand(__inline)` task. This means that if the user wants it
+to execute it in a separate task, they must wrap the task themselves
```
Missing period at end of sentence.

arjunkunna replied:
Added

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,237 @@
 
-**Important:** like `make_plan`, `make_plan_distrib` overwrites the
-input and output regions `r` and `s`.
+#### 3. Execute the plan
```
Title Case?

arjunkunna replied:
Edited

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,237 @@
-because the plans must be used by the host-side code to launch the FFT
-kernels. Therefore, when `execute_plan_task` is mapped onto the GPU it
-is very important to map the `p` region into zero-copy memory.
+In this case, the user creates a 3D interface:
```
Well it wouldn't be a 3D interface if the transform is 256 x 256?

arjunkunna replied:
I've modified the language here a bit: "Since the transform is a 2D one, the user creates a interface with `itype` of dimension N+1: in this case, an `int3d`. The last dimension is used to store the number of batches."

I didn't think there was an easy way to get around the fact that they have to use an `itype` of size `N+1` if we want to use the same API / interface -- the last dimension is the no. of batches -- but let me know if you feel otherwise

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,237 @@
+
+Please also refer to the `test3d_batch` and `test3d_batch_real` examples in `fft_test.rg` for reference.
+
+Batched transforms are supported on both CPU and GPUs, for 1 and 2 dimensions. For GPUs, both real-to-complex and complex-to-complex tranforms are supported (for both `complex32` and `complex64`). For CPU, only `complex64`-to-`complex64` transforms are supported currently.
```
And 3 dimensions?

arjunkunna replied:
Updated - we now support 3 dimensions and require Regent to be build with the MAX_DIM = 4 flag

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,237 @@
-the node where the plan was originally created.
+Next items in the pipeline include batch transforms, as well as distributed
+transforms across multiple nodes. Please let us know if there are specific
+features that may be helpful.
```
Batch transforms are done now?

And same note about using issues for feature requests.

arjunkunna replied:
Edited

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,237 @@
+
+## Future Developments
 
-**Important:** like `make_plan_distrib`, the index launch issued by
-`destroy_plan_distrib` must be mapped so that each point task runs on
-the node where the plan was originally created.
```
Seems like this didn't make it into the final version? Also I'm not sure if the new README has any discussion of distrib plans and their implications at all.

arjunkunna replied:
Item 3 on my todo list

arjunkunna replied:
Added commentary on distrib plans and added this comment back in too

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,237 @@
-  * FFT operations are backed by FFTW, so at the moment the run
-    CPU-only and only on a single node. (OpenMP is hypothetically
-    possible, but I haven't tested it yet.)
+## Version History
```
I'd move this into a separate file called `CHANGES.md`.

arjunkunna replied:
Done

elliottslaughter commented on legion:
I'd rather not use submodules.

For end-users, pinning Legion like this is appropriate. For libraries, we really shouldn't specify a specific pinned Legion. We should aim to use what the user provides; and to the extent that there are bounds or requirements we can specify those.

arjunkunna replied:
This makes sense. Removed the legion module

elliottslaughter commented on test.sh:
```diff
@@ -0,0 +1,54 @@
+sudo apt-get update -qq
+sudo apt-get install -qq cuda-toolkit-12-2
+sudo apt-get install nvidia-utils-515
+sudo apt install -y libjemalloc2 # valgrind
```
Can we re-evaluate which of these options we need? I am 99.99% sure we do not need the driver. There may be a different set of packages that include the subset of things we need here.

arjunkunna replied:
I removed the driver and cleaned this up a little. Let me know if you think it can be slimmed down further

elliottslaughter commented on test.sh:
```diff
@@ -0,0 +1,54 @@
+export PATH="$CUDA_PATH/bin:$PATH"
+export LD_LIBRARY_PATH="$CUDA_PATH/lib:$LD_LIBRARY_PATH"
+export LD_LIBRARY_PATH="$CUDA_PATH/lib64:$LD_LIBRARY_PATH"
+export LD_LIBRARY_PATH="/usr/lib/x86_64-linux-gnu:$LD_LIBRARY_PATH"
```
This seems redundant?

arjunkunna replied:
Cleaned this up

elliottslaughter commented on test.sh:
```diff
@@ -0,0 +1,54 @@
+#ls -lH /usr/lib/x86_64-linux-gnu
+
+#find /usr/ -name 'libcuda.so.*'
+#nvidia-smi
```
Let's clean up this debugging type stuff.

arjunkunna replied:
Cleaned up

elliottslaughter commented on test.sh:
```diff
@@ -0,0 +1,54 @@
+
+nvcc --version
+git clone --depth 1 --branch stable https://github.com/StanfordLegion/legion.git
+CC=gcc CXX=g++ DEBUG=${DEBUG:-0} USE_GASNET=0 USE_CUDA=1 ./legion/language/scripts/setup_env.py --terra-binary
```
`--terra-binary` is the default option. `DEBUG=0` is the default option so you don't need to specify it.

arjunkunna replied:
Edited

elliottslaughter commented on test.sh:
```diff
@@ -0,0 +1,54 @@
+CC=gcc CXX=g++ DEBUG=${DEBUG:-0} USE_GASNET=0 USE_CUDA=1 ./legion/language/scripts/setup_env.py --terra-binary
+./install.py
+#LD_PRELOAD="$LIBJEMALLOC_SO_PATH" ./legion/language/regent.py test/fft_test.rg -fgpu cuda -fgpu-offline 1 -fgpu-arch pascal
+./legion/language/regent.py test/fft_test.rg -fgpu cuda -fgpu-offline 1 -fgpu-arch pascal
```
In this case, we are not actually running the tests at all (because we have no GPU).

I appreciate at least confirming that the GPU code builds, but I think we need a CPU test so that we know the software actually works.

arjunkunna replied:
Added `./legion/language/regent.py test/fft_test.rg -fgpu none` as we discussed. Also separated code paths (i.e., adding _gpu functions for `execute` and `destroy` so that CPU and GPU can run separately. I believe we now test CPU code execution and build GPU code.

elliottslaughter commented on test/test_mapper.cc:
```diff
@@ -24,85 +27,33 @@ using namespace Legion::Mapping;
-                                             Processor target_proc,
-                                             const RegionRequirement &req);
+  FFTTestMapper(MapperRuntime *rt, Machine machine, Processor local, const char *mapper_name);
+  virtual Memory default_policy_select_target_memory(MapperContext ctx, Processor target_proc, const RegionRequirement &req, MemoryConstraint mc = MemoryConstraint());
```
Let's use `override` instead of `virtual`.

arjunkunna replied:
Edited

elliottslaughter commented on src/fft.rg:
```diff
@@ -1,86 +1,83 @@
--- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
--- See the License for the specific language governing permissions and
--- limitations under the License.
-
```
Definitely should not delete the license header. You can add your name to the list of copyright holders since you are doing this work on your own. Also please update the copyright year.

arjunkunna replied:
Added this back in and edited year

elliottslaughter commented on src/fft.rg:
```diff
@@ -1,86 +1,83 @@
--- See the License for the specific language governing permissions and
--- limitations under the License.
-
---[[--
-Regent FFT library.
-]]
```
Another important comment we should not be deleting.

arjunkunna replied:
Added this back

elliottslaughter commented on src/fft.rg:
```diff
@@ -1,86 +1,83 @@
-
 import "regent"
 
+--Import libraries
```
Self-evident?

arjunkunna replied:
Removed this comment

elliottslaughter commented on src/fft.rg:
```diff
@@ -1,86 +1,83 @@
+local launcher = require("std/launcher")
 local data = require("common/data")
-local cudahelper = require("regent/cudahelper")
+local cmapper = require("test_mapper")
```
I think we should probably move this (and the launcher) into the test program? This is a library and should work whatever way the user wants it.

arjunkunna replied:
Moved to fft_test.rg

elliottslaughter commented on src/fft.rg:
```diff
@@ -1,86 +1,83 @@
+--Import C and FFTW APIs
 local c = regentlib.c
 local fftw_c = terralib.includec("fftw3.h")
 terralib.linklibrary("libfftw3.so")
```
This should be `regentlib.linklibrary` to support offline mode.

arjunkunna replied:
Fixed. Also changed the line below `regentlib.linklibrary("libcufft.so")` 

I kept the `terralib.includec()s` though, I couldn't find a change for that

elliottslaughter commented on src/fft.rg:
```diff
@@ -1,86 +1,83 @@
 local c = regentlib.c
 local fftw_c = terralib.includec("fftw3.h")
 terralib.linklibrary("libfftw3.so")
+--terralib.linklibrary("libfftw3f.so")
```
In general let's clean up comments like this. If this is indicative of a code path we want, let's add it as a specific option. Otherwise, if this is unnecessary, we can just remove it.

arjunkunna replied:
Added this back now that we are supporting single-precision FFTW

elliottslaughter commented on src/fft.rg:
```diff
@@ -1,86 +1,83 @@
   assert(regentlib.is_index_type(itype), "requires an index type as the first argument")
   local dim = itype.dim
-  assert(dim >= 1 and dim <= 3, "currently only 1 <= dim <= 3 is supported")
-  assert(dtype == complex64, "currently only complex64 is supported")
```
Looks like this check went missing. We may support more types now, but we still need to check them.

elliottslaughter replied:
I think I'd generally prefer that the asserts go as early as possible in this file, like was originally written. These are basically sanity checks on the user input. Because they come from the user, we should not assume they are correct. Therefore, before we go do anything else, let's check all those things, and then we can go set up the rest of the file.

arjunkunna replied:
Edited to cover all options and moved up

elliottslaughter commented on src/fft.rg:
```diff
@@ -1,86 +1,83 @@
 
-function fft.generate_fft_interface(itype, dtype)
+-- Create the FFT interface
+--itype should be the index type of the transform (int1d for 1d/int2d for 2d). dtype_in = complex64/complex32/real/float; dtype_out = complex32/complex64 depending on transform. 
```
We should aim to make this usable for documentation. You can see the format for comments here:

https://gitlab.com/StanfordLegion/legion/-/blob/master/language/src/std/format.rg#L300-304

arjunkunna replied:
Added comment in proper format here. I also added similar documentation for all of the tasks exposed to users

elliottslaughter commented on src/fft.rg:
```diff
@@ -1,86 +1,83 @@
+  if gpu_available then 
     fspace iface_plan {
       p : fftw_c.fftw_plan,
+      float_p : fftw_c.fftwf_plan,
```
We'll only ever be doing one of double or float at a time, right? Then this feels like it ought to be an if statement.

arjunkunna replied:
I tried separating this, but this had trouble compiling - i think because [these lines 
](https://github.com/arjunkunna/regent-fft/blob/6d185fa93566ceac03d9ec6b0ff25d838eb94576/src/fft.rg#L307)

require both `p.float_p` and `p.p` to be present. I guess I could maybe also separate them (into different tasks?) like we do with `if gpu_available` but that seemed to be adding complexity and wasn't sure if it was worth the savings. Happy to do that though.

elliottslaughter commented on src/fft.rg:
```diff
@@ -1,86 +1,83 @@
     }
   end
+  
+  --Store plan fspace in our interface
```
Seems self-evident?

arjunkunna replied:
Removed comment

elliottslaughter commented on src/fft.rg:
```diff
@@ -1,86 +1,83 @@
-                 physical : c.legion_physical_region_t,
-                 field : c.legion_field_id_t)
+    --Function to get base pointer of region: returns base_pointer
+    local terra get_base(rect : rect_t, physical : c.legion_physical_region_t, field : c.legion_field_id_t)
```
Did this actually change? If not I don't think we need to reformat this.

elliottslaughter replied:
Also the comment seems to be mostly mirroring what is in the code itself; I think comments are helpful in explaining details you wouldn't understand by simply reading what the code says.

arjunkunna replied:
Returned the formatting to the original version with arguments on different lines

elliottslaughter commented on src/fft.rg:
```diff
@@ -93,27 +90,42 @@ function fft.generate_fft_interface(itype, dtype)
       end
-      regentlib.assert(offsets[0].offset == terralib.sizeof(t), "stride does not match expected value")
 
+      --regentlib.assert(offsets[0].offset == terralib.sizeof(complex64), "stride does not match expected value")
```
Definitely should not remove this without an explanation.

arjunkunna replied:
Added this back

elliottslaughter commented on src/fft.rg:
```diff
@@ -93,27 +90,42 @@ function fft.generate_fft_interface(itype, dtype)
 
+      --regentlib.assert(offsets[0].offset == terralib.sizeof(complex64), "stride does not match expected value")
       destroy_accessor(accessor)
-
```
As a general rule let's try to avoid diffs like these.

arjunkunna replied:
Maintained this diff

arjunkunna replied:
Should be resolved

elliottslaughter commented on src/fft.rg:
```diff
@@ -93,27 +90,42 @@ function fft.generate_fft_interface(itype, dtype)
+      return offsets
+    end
+
+    return rect_t, get_base, get_offset
```
I don't think we should make this a separate function. If we need the offset out of the other one, we can make it return a struct containing the base pointer and offset.

arjunkunna replied:
Combined into `get_base` and created a struct

elliottslaughter commented on src/fft.rg:
```diff
@@ -93,27 +90,42 @@ function fft.generate_fft_interface(itype, dtype)
-  -- Hack: need to retrieve context without __context() to circumvent leaf checker here.
+  --Define get_base functions for input, output, and plan regions
+  local rect_plan_t, get_base_plan, get_offset_plan = make_get_base(1, iface.plan) --get_base_plan returns a base_pointer to a region with fspace iface.plan. (always dim = 1 because plan regions are dim 1: 'var p = region(ispace(int1d, 1), fft1d.plan)')
+  local rect_t_in, get_base_in, get_offset_in = make_get_base(dim, dtype_in) --get_base returns a base pointer to a region with fspace dtype
+  local rect_t_out, get_base_out, get_offset_out = make_get_base(dim, dtype_out) --get_base returns a base pointer to a region with fspace dtype
```
The `_t` suffix is for type; this is a convention from C. It always goes last.

elliottslaughter replied:
And again, let's avoid comments that just restating what the code is doing.

arjunkunna replied:
Removed comments, switched order of variable name so that `_t` is at the end

elliottslaughter commented on src/fft.rg:
```diff
@@ -93,27 +90,42 @@ function fft.generate_fft_interface(itype, dtype)
 
-  -- Hack: need to retrieve context without __context() to circumvent leaf checker here.
+  --Define get_base functions for input, output, and plan regions
+  local rect_plan_t, get_base_plan, get_offset_plan = make_get_base(1, iface.plan) --get_base_plan returns a base_pointer to a region with fspace iface.plan. (always dim = 1 because plan regions are dim 1: 'var p = region(ispace(int1d, 1), fft1d.plan)')
```
We should probably take a separate `plan_dim` as an input from the user because it's really the user's choice how many dimensions they want.

arjunkunna replied:
As discussed over email, keeping this at 1 for now.

elliottslaughter commented on src/fft.rg:
```diff
@@ -93,27 +90,42 @@ function fft.generate_fft_interface(itype, dtype)
+  local rect_t_out, get_base_out, get_offset_out = make_get_base(dim, dtype_out) --get_base returns a base pointer to a region with fspace dtype
+
+
+  -- Takes a c.legion_runtime_t and returns c.legion_runtime_get_executing_processor(runtime, ctx)
```
A more helpful comment would be something like "used to branch on CPU vs GPU execution inside of tasks". Again, don't summarize what the code does. Tell us why the code is doing that.

arjunkunna replied:
Edited this comment

elliottslaughter commented on src/fft.rg:
```diff
@@ -93,27 +90,42 @@ function fft.generate_fft_interface(itype, dtype)
   end
 
-  -- FIXME: Keep this in sync with default_mapper.h
+  -- Note from Elliott: Keep this in sync with default_mapper.h
```
It is from me but is something that applies universally when we have constants to keep in sync. 😃

arjunkunna replied:
Haha. Returned to original

elliottslaughter commented on src/fft.rg:
```diff
@@ -126,16 +138,12 @@ function fft.generate_fft_interface(itype, dtype)
-    -- result in a double destroy if we free here.
-
-    -- c.legion_future_destroy(f)
+    -- c.legion_future_destroy(f) -- Comment from Elliott: I thought Regent was supposed to copy on assignment, but that seems not to happen here, so this would result in a double destroy if we free here.
```
Seems like this is being reformatted for no purpose?

arjunkunna replied:
Returned to original formatting

elliottslaughter commented on src/fft.rg:
```diff
@@ -144,133 +152,435 @@ function fft.generate_fft_interface(itype, dtype)
     return iface.get_tunable(DEFAULT_TUNABLE_NODE_COUNT)
   end
 
+  -- Get the number of GPUs
```
Remove. This is entirely specified by the name of the task, and in fact the task name here is actually *more* informative than the comment because it specifies that these are *local* GPUs.

arjunkunna replied:
Removed

elliottslaughter commented on src/fft.rg:
```diff
@@ -144,133 +152,435 @@ function fft.generate_fft_interface(itype, dtype)
   end
 
-  __demand(__inline)
+  --Task to return pointer to plan. Takes plan region and returns pointer to plan. Calls get_base_plan made above.
+   __demand(__inline)
```
I think this is not part of the public interface? To me, this comment seems to recapitulate the task signature, which makes it non-helpful. If it were a public interface, we'd want to document it (ideally with a description that is more helpful), but I think we just don't need to do that at all in this case.

arjunkunna replied:
Removed this, along with any other similar comments for non-public tasks

elliottslaughter commented on src/fft.rg:
```diff
@@ -144,133 +152,435 @@ function fft.generate_fft_interface(itype, dtype)
   where reads(plan) do
-    -- Hack: 3Bneed to use raw access to circument CUDA checker here.
-    var pr = __physical(plan)[0]
+    format.println("In get_plan...")
```
This section of code has a bunch of comments that seem to be summarizing basic language features. I would remove all comments that either explain how Regent features work (if this is not obvious, the place to update is the Regent language reference), and also remove all comments that summarize or restate code without adding detail. I'm on the fence about "section" comments that mark out large sections of the code; they can be helpful but I don't think this particular function is complicated enough to warrant it?

After you do that, please double check the diff and avoid reformatting things that don't need to be reformatted. Ideally there should be a diff IFF the code actually changed.

arjunkunna replied:
Removed the regent-feature comments. Tried to clean up comments and diff. Let me know if there's still work to be done on this.

elliottslaughter commented on src/fft.rg:
```diff
@@ -144,133 +152,435 @@ function fft.generate_fft_interface(itype, dtype)
+        --Call cufftPlanMany: cufftResult cufftPlanMany(cufftHandle *plan, int rank, int *n, int *inembed, int istride, int idist, int *onembed, int ostride, int odist, cufftType type, int batch) --rank = dimensionality of transform (1,2,3)
+        format.println("Calling cufftPlanMany...")
+
+        var ok = 0
```
Instead of having an `ok` variable, I would make a `check_cufft` function or macro that checks the result, so you can write something like:

```
check_cufft(cufft_c.cufftPlanMany(...))
```

You can then write `check_cufft` to give you nice, readable error messages with string representations of the error codes and so on.

arjunkunna replied:
Created a 'cufft_assert' function

elliottslaughter commented on src/fft.rg:
```diff
@@ -144,133 +152,435 @@ function fft.generate_fft_interface(itype, dtype)
+
+        var ok = 0
+
+        if dtype_size == 8 and real_flag then
```
I think it would be more intuitive to just write `dtype_in == float` or something?

arjunkunna replied:
Created bools at the start of `generate_fft_interface` to identify which case we are in

elliottslaughter commented on src/fft.rg:
```diff
@@ -144,133 +152,435 @@ function fft.generate_fft_interface(itype, dtype)
+      
+      --GPU not identified: return error
+      else 
+        format.println("GPU processor not identified: TOC_PROC not equal to processor kind")
```
Redundant, we're about to assert.

arjunkunna replied:
Removed this

elliottslaughter commented on src/fft.rg:
```diff
@@ -144,133 +152,435 @@ function fft.generate_fft_interface(itype, dtype)
-  task iface.make_plan(input : region(ispace(itype), dtype),
-                       output : region(ispace(itype), dtype),
-                       plan : region(ispace(int1d), iface.plan))
+  --CPU version of make_plan. Calls make_plan_gpu if necessary
```
Actually it's not the CPU version of `make_plan` because it is the user's entry point to both CPU and GPU functionality.

This is where it would be a good idea to write a proper documentation string.

arjunkunna replied:
Wrote documentation string, edited description

elliottslaughter commented on src/fft.rg:
```diff
@@ -144,133 +152,435 @@ function fft.generate_fft_interface(itype, dtype)
+      var n : int[dim]
+      ;[data.range(dim):map(function(i) return rquote n[i] = hi.x[i] - lo.x[i] + 1 end end)] --n is a array of size dim storing the size of each dimension
+      format.println("calling fftwf_plan_dft_r2c")
+      --p.float_p = fftw_c.fftwf_plan_dft_r2c(dim, &n[0], [&float](input_base), [&fftw_c.fftwf_complex](output_base), fftw_c.FFTW_ESTIMATE)
```
We don't actually call the function? Why is this code path even here?

arjunkunna replied:
This is resolved now that we handle single-precision FFTW

elliottslaughter commented on src/fft.rg:
```diff
@@ -144,133 +152,435 @@ function fft.generate_fft_interface(itype, dtype)
+    if dtype_size == 8 and real_flag then
+      format.println("R2C: Float to Complex32: Not Supported")
+      --AK Note: I don't actually understand this that well, how the size of each dimension is stored in the 'n' array
+      var n : int[dim]
+      ;[data.range(dim):map(function(i) return rquote n[i] = hi.x[i] - lo.x[i] + 1 end end)] --n is a array of size dim storing the size of each dimension
```
It seems like there's a bunch of redundant code like this. In general, when we don't need to duplicate code N times, it's better to avoid that because it makes everything else harder to understand.

arjunkunna replied:
Pulled this out of the for-loops

elliottslaughter commented on src/fft.rg:
```diff
@@ -144,133 +152,435 @@ function fft.generate_fft_interface(itype, dtype)
+  if gpu_available then
+    __demand(__cuda, __leaf)
+
+    task make_plan_gpu_batch(input : region(ispace(itype), dtype_in), output : region(ispace(itype), dtype_out), plan : region(ispace(int1d), iface.plan), address_space : c.legion_address_space_t)
```
So, I think this is actually not correct? We need an `itype` in N+1 dimensions.

arjunkunna replied:
I think this is correct, the interface reads the extra batch dimension as part of the itype dimension count. But to the user, they don't have to worry about any of that. That's why we have the `n_batch` variable to only capture the dimensions we want. Let me know if you prefer we handle this differently though

elliottslaughter commented on src/fft.rg:
```diff
@@ -144,133 +152,435 @@ function fft.generate_fft_interface(itype, dtype)
+        --rank[In] – How many dimensions a single transform has: 2
+        --n[In] - Array of size rank, describing the size of each dimension: {256, 256}
+        --inembed[In] – Array of size rank. Here we pass the real dimensions of the input array. It should contain the dimension size + padding: {256, 256} again, because input is not padded.
+        --istride[In] – Stride between two consecutive elements in lowest dimension: 1
```
Stride isn't actually guaranteed to be 1, is it?

arjunkunna replied:
Still looking into this

arjunkunna replied:
Created an `istride` variable to calculate the actual stride. I'm assuming the `istride` and `ostride` are identical here.

elliottslaughter commented on src/fft.rg:
```diff
@@ -144,133 +152,435 @@ function fft.generate_fft_interface(itype, dtype)
+    end
+
+    --Call fftw_plan fftw_plan_many_dft(int rank, const int *n, int howmany, fftw_complex *in, const int *inembed, int istride, int idist, fftw_complex *out, const int *onembed, int ostride, int odist, int sign, unsigned flags)
+    --Taking a transform where {256x256x7} is passed (i.e. 7 batches of 256x256), the parameters should be as follows:
+      --rank[In] – How many dimensions a single transform has: 2
+      --n[In] - Array of size rank, describing the size of each dimension: {256, 256}
+      --inembed[In] – Array of size rank. Here we pass the real dimensions of the input array. It should contain the dimension size + padding: {256, 256} again, because input is not padded.
+      --istride[In] – Stride between two consecutive elements in lowest dimension: 1
+      --idist[In] – Distance between two input batches: 256 * 256
+      --onembed[In] – Array of size rank. Here we pass the real dimensions of the output array. It should contain the dimension size + padding: {256, 256} again, because input is not padded.
+      --ostride[In] – Stride between two consecutive elements in lowest dimension: 1
+      --odist[In] – Distance between two output batches: 256 * 256
+      --type[In] – The transform data type: Z2Z or D2Z.
+      --batch[In] – How many batches should be computed: 7
```
Is this accurate? I don't think this matches the FFTW interface.

arjunkunna replied:
You're right -- this was copied from CUFF. Have removed this.

elliottslaughter commented on src/fft.rg:
```diff
@@ -144,133 +152,435 @@ function fft.generate_fft_interface(itype, dtype)
+
+    end
     p.address_space = address_space
-    ;[(function()
-         if use_cuda then
-           return rquote
-             if iface.get_num_local_gpus() > 0 then
-               p.cufft_p = make_plan_gpu(input, output, address_space)
-             end
-           end
-         else
-           return rquote end
-         end
-       end)()]
```
Are we missing the CUDA case?

arjunkunna replied:
Handled by `if gpu_available` so this is resolved

elliottslaughter commented on src/fft.rg:
```diff
@@ -144,133 +152,435 @@ function fft.generate_fft_interface(itype, dtype)
-    p.address_space
-    ;[(function() if use_cuda then return rquote p.cufft_p = 0 end else return rquote end end end)()]
+
+    if gpu_available then
+      p.cufft_p = 0
+    end
```
I think you can't just do this. The field does not even exist in the non-GPU case.

arjunkunna replied:
Resolved by adding switch to `rescape` / `remit rquote` and then breaking into a separate `execute_plan_gpu` -- let me know if this doesn't fix this

elliottslaughter commented on src/fft.rg:
```diff
@@ -279,48 +589,133 @@ function fft.generate_fft_interface(itype, dtype)
+    --If in GPU mode, use cufftExec
     if c.legion_processor_kind(proc) == c.TOC_PROC then
       c.printf("execute plan via cuFFT\n")
+      var ok = 0
```
Should we split this out to have a separate GPU task like the others?

arjunkunna replied:
Yes, have split this out.

elliottslaughter commented on src/fft.rg:
```diff
@@ -279,48 +589,133 @@ function fft.generate_fft_interface(itype, dtype)
   end
 
+
+  --execute_plan tasks are demand_(__inline tasks). This means that if the user wants it to execute it in a separate task, they must wrap the task themselves -- this is what execute_plan_task does.
```
Let's write some better documentation for this.

arjunkunna replied:
Have tried to make clear what the purpose of the inlined  vs. non_inlined tasks are -- let me know if this isn't clear still

elliottslaughter commented on src/fft.rg:
```diff
@@ -279,48 +589,133 @@ function fft.generate_fft_interface(itype, dtype)
+    var proc = get_executing_processor(__runtime())
+
+    -- If using GPUs, call cufftDestroy
+    if c.legion_processor_kind(proc) == c.TOC_PROC then
+      c.printf("Destroy plan via cuFFT\n")
+      --Function: cufftResult cufftDestroy(cufftHandle plan)
+      cufft_c.cufftDestroy(p.cufft_p) 
```
Same comment as before.

arjunkunna replied:
Should be resolved by split into _gpu task

elliottslaughter commented on test/fft_test.rg:
```diff
@@ -1,108 +1,409 @@
--- Copyright 2020 Stanford University
---
--- Licensed under the Apache License, Version 2.0 (the "License");
--- you may not use this file except in compliance with the License.
--- You may obtain a copy of the License at
---
---     http://www.apache.org/licenses/LICENSE-2.0
---
--- Unless required by applicable law or agreed to in writing, software
--- distributed under the License is distributed on an "AS IS" BASIS,
--- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
--- See the License for the specific language governing permissions and
--- limitations under the License.
-
```
Again, do not remove license headers: put it back, add your name, and update the year.

arjunkunna replied:
Added back

elliottslaughter commented on test/fft_test.rg:
```diff
@@ -1,108 +1,409 @@
+local format = require("std/format")
+
+--Import cuFFT API
+local cufft_c = terralib.includec("cufftXt.h")
+terralib.linklibrary("libcufft.so")
```
Why does the user need this at all?

arjunkunna replied:
Removed these

elliottslaughter commented on test/fft_test.rg:
```diff
@@ -1,108 +1,409 @@
+__demand(__inline, __leaf)
+task print_array_double_complex(input : region(ispace(int1d), complex64), arrayName: rawstring)
+where reads (input) do
+  format.println("\n{} = [", arrayName)
```
You can `format.println("")` to get just a newline.

It wasn't clear to me below if you really wanted as many newlines as you've got here. If you want to NOT put in a newline you can use `format.print()`.

arjunkunna replied:
Thanks -- have made the modifications where applicable

elliottslaughter commented on test/fft_test.rg:
```diff
@@ -1,108 +1,409 @@
+
+-- Task to print out input or output array with fieldspace complex32. Takes a region and a string representing the name of the array
+__demand(__inline, __leaf)
+task print_array_float_complex(input : region(ispace(int1d), complex32), arrayName: rawstring)
```
Overall this seems like a lot of redundant code that we could unify through some metaprogramming.

elliottslaughter replied:
And if we need to support complex types in `format`, let's do that.

arjunkunna replied:
Consolidated a bunch of tasks with metaprogramming. Also submitted MR to legion for `format`.

elliottslaughter commented on test/fft_test.rg:
```diff
@@ -1,108 +1,409 @@
+  format.println("]\n")
+end
+
+--task compare_regions_double_complex(r1 : region(ispace(int1d), complex64), r2 : region(ispace(int1d), complex64))
```
We don't actually test anything?

arjunkunna replied:
Removed this. Have a `make_compare_regions_task` through metaprogramming now.

elliottslaughter commented on test/fft_test.rg:
```diff
@@ -1,108 +1,409 @@
+----SET UP INTERFACES----
+
+--Usage: function fft.generate_fft_interface(itype, dtype): itype = int1d, dtype = complex64, dim = itype.dim =1
+local fft1d = fft.generate_fft_interface(int1d, complex64, complex64)
+local fft2d = fft.generate_fft_interface(int2d, complex64, complex64)
+local fft3d = fft.generate_fft_interface(int3d, complex64, complex64)
```
Better to use some consistent naming below so that we can use fewer comments.

e.g. `fft1d_complex32`, `fft2d_float`, etc.

arjunkunna replied:
Now use the naming convention of `fftXd_<input type>_<output type>`

elliottslaughter reviewed (commented):
Thanks for working on this. A bunch of detailed comments are inline. Overall, I think there are a couple things that stand out to me:

 * We need to test CPU execution in our CI, since that is the only mode that we can actually run. (We can also build the GPU code, but since we will never be able to run, it's not something we can rely on.)
 * The test code should actually verify the output if possible.
 * Overall there is a lot of formatting clean up to do to make things clear and obvious and minimize diffs where no actual changes occurred.

elliottslaughter commented:
Also, could you run clang-format on the C++ code?

You can find a script here:

 * script: https://github.com/StanfordLegion/fuzzer/blob/master/format.sh
 * config: https://github.com/StanfordLegion/fuzzer/blob/master/.clang-format
 * example job: https://github.com/StanfordLegion/fuzzer/blob/master/.github/workflows/main.yml#L9-L14

arjunkunna committed:
Update REAMDE

arjunkunna committed:
Add .clang-format and update formatting of .h/.cc files

arjunkunna committed:
Update header guards

arjunkunna committed:
Clean redundant stuff

arjunkunna committed:
Run clang-format

arjunkunna committed:
Keep note; move version history to CHANGED.md [skip ci]

arjunkunna committed:
Use override; remove virtual [skip ci]

arjunkunna committed:
Remove legion

arjunkunna committed:
Format comments; remove redundant ones

arjunkunna committed:
Remove redundant output in favor of assertion [skip ci]

arjunkunna committed:
Remove redudnant comments; wrap long comments to 80 column

arjunkunna committed:
Fix naming convention; remove redundant comments

arjunkunna committed:
Add documentation comments [skip ci]

arjunkunna committed:
Update doc comments [skip ci]

arjunkunna committed:
Remove workflow_dispatch (manual trigger) [skip ci]

arjunkunna committed:
Update copyright year; remove redundant comments [skip ci]

arjunkunna committed:
Add missing annotations for print tasks

arjunkunna committed:
Verify tests' outputs

arjunkunna committed:
Remove redundant variables [skip ci]

Commit:
Readme update

Commit:
cleaning up

Commit:
Merge branch 'master' of https://github.com/arjunkunna/regent-fft

Commit:
cleaning up

Commit:
Merge branch 'master' of https://github.com/arjunkunna/regent-fft

Commit:
more cleanup

Commit:
add extra line in test

Commit:
rescape issues

Commit:
Revert "rescape issues"

This reverts commit d1d11998de2b0e4a7065b03fbc3a866e67c66583.

Commit:
if gpu_available edit

Commit:
move launched to fft_test.rg

Commit:
commit before reog of make plan execute

Commit:
fgpu none works, distrib broken

Commit:
try debug mode

Commit:
try debug mode

Commit:
try debug mode

Commit:
try debug mode

Commit:
remove legion

Commit:
modigy gitginore

Commit:
modigy gitginore

Commit:
change to regentlib.linklibrary

arjunkunna committed:
Update README.md

Commit:
include asserts

Commit:
Merge branch 'master' of https://github.com/arjunkunna/regent-fft

Commit:
cleanup

arjunkunna committed:
Update README.md

Commit:
added offset assert back

Commit:
Merge branch 'master' of https://github.com/arjunkunna/regent-fft

arjunkunna committed:
Use metaprogramming for print (float and double) and comparison functions

arjunkunna committed:
Merge branch 'master' of https://github.com/arjunkunna/regent-fft

Commit:
add comments on __inline vs. _task tasks

Commit:
Merge branch 'master' of https://github.com/arjunkunna/regent-fft

arjunkunna committed:
Update README.md

arjunkunna committed:
Update README.md

arjunkunna committed:
Update README.md

arjunkunna committed:
Update README.md

Commit:
add check_cufft and get_offset

Commit:
change batch terminology

arjunkunna committed:
Update README.md

Commit:
minor comments

Commit:
Merge branch 'master' of https://github.com/arjunkunna/regent-fft

Commit:
test int4d

Commit:
int4d attempt but failing test

Commit:
int4d working now

Commit:
enable FFTW single-precision, also dtype_flow

Commit:
task descriptions

Commit:
accepting multiple strides

Commit:
cleanup

Commit:
comment out complex print

Commit:
cleanup

arjunkunna committed:
Rearrange tests: single-precision, double-precision, and then double-precision batch transforms

arjunkunna committed:
Minor cleanup: remove trailing white space, format multiline comments to wrap at column 80, and add missing EOL

arjunkunna committed:
Remove redundant require [skip ci]

arjunkunna committed:
Update README.md

arjunkunna committed:
Update README.md

arjunkunna committed:
Update README.md

Commit:
small nits

Commit:
small nits

arjunkunna committed:
Update CHANGES.md

arjunkunna committed:
Minor updates

arjunkunna commented:
Two additional notes: 

-  I ultimately found that I was able to simultaneously use both single- and double- precision of FFTW by using the `enable-single` flag. Here is the edit to `install.py`: https://github.com/arjunkunna/regent-fft/blob/ec4fd76a137012446f0f827b772ac445256d6cac/install.py#L55C135-L55C150

- I also removed the unneeded print tasks in `fft_test.rg` now that we have `complex` in `format.rg` and can use metaprogramming ([this ](https://github.com/StanfordLegion/regent-fft/pull/2/commits/de9b212e2b48c1d29717eba98a848df25a42afae)commit). I switched to using the gitlab master branch of legion (was previously using the github stable version)

Commit:
add complex to format.rg

Commit:
Merge branch 'master' of https://github.com/arjunkunna/regent-fft

elliottslaughter commented on .github/workflows/master.yml:
```diff
@@ -0,0 +1,30 @@
+        run: |
+          clang-format -i test/*.{h,cc}
+          git status
+          git diff
```
I think you need the `--exit-code` flag in order to make this do what you want. Otherwise I'm not actually sure what `git diff` returns here.

arjunkunna replied:
Added --exit-code so now it fails if there are diffs. https://github.com/arjunkunna/regent-fft/commit/89c88e332916e7ffd99d9d67b8863b7f5b6b6e8d

elliottslaughter commented on .gitignore:
```diff
@@ -1,4 +1,9 @@
 env.sh
+valgrind*
+gdb*
+*.txt
```
Why are we ignoring `.txt` files?

arjunkunna replied:
Removed that line

elliottslaughter commented on .gitignore:
```diff
@@ -1,4 +1,9 @@
-
-/fftw-*
+legion
+fftw-*
```
Is there a reason to remove the `/`?

arjunkunna replied:
Added this back

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
 
-## Installation
+[![Regent-FFT
+Tests](https://github.com/StanfordLegion/regent-fft/actions/workflows/main.yml/badge.svg)](https://github.com/StanfordLegion/regent-fft/actions/workflows/master.yml)
```
```suggestion
Tests](https://github.com/StanfordLegion/regent-fft/actions/workflows/main.yml/badge.svg)](https://github.com/StanfordLegion/regent-fft/actions)
```

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
+[![Regent-FFT
+Tests](https://github.com/StanfordLegion/regent-fft/actions/workflows/main.yml/badge.svg)](https://github.com/StanfordLegion/regent-fft/actions/workflows/master.yml)
 
+This is a fast fourier transform library built in Regent.
```
```suggestion
This is a fast fourier transform (FFT) library built in Regent.
```

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
+## Description
+
+At a high level, the library takes the input matrix for the Discrete Fourier
+Transform in the form of a region, and saves the output in an output region.
```
```suggestion
Regent-FFT takes the input matrix for the discrete fourier
transform (DFT) in a region, and saves the output in an output region.
```

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
+Transform in the form of a region, and saves the output in an output region.
+
+The library currently supports transforms up to 3 dimensions, and can be
+configured to run on either a CPU or a GPU.
```
```suggestion
configured to run on either CPUs or GPUs.
```

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
+### Install
+
+First, make sure you have Regent installed. (If you are using Sapling, please
+refer to the [Sapling guide](https://github.com/StanfordLegion/sapling-guide)).
```
```suggestion
refer to the [Sapling guide](https://github.com/StanfordLegion/sapling-guide).)
```

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
+
+First, make sure you have Regent installed. (If you are using Sapling, please
+refer to the [Sapling guide](https://github.com/StanfordLegion/sapling-guide)).
+Be sure to install with `MAX_DIM=4` flag enabled.
```
```suggestion
Be sure to install with `MAX_DIM=4` (for Make) or `-DLegion_MAX_DIM=4` (for CMake).
```

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
+refer to the [Sapling guide](https://github.com/StanfordLegion/sapling-guide)).
+Be sure to install with `MAX_DIM=4` flag enabled.
+
+Note that you will want to load the following modules instead:
```
```suggestion
Note that on Sapling you will want to load the following modules:
```

elliottslaughter commented on README.md:
````diff
@@ -1,189 +1,337 @@
+module load slurm mpi cmake cuda llvm
 ```
+
+Otherwise, refer to the Regent installation instructions
+[here](https://regent-lang.org/install/) or the Legion ones
+[here](https://legion.stanford.edu/starting/).
````
This seems a bit redundant with the above? I think we only need to mention installation once.

arjunkunna replied:
Removed this.

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
+srun -n 1 -N 1 -c 40 -p gpu --exclusive --pty bash --login
+<navigate to your .rg file>
+source env.sh
+../legion/language/regent.py test/fft_test.rg
```
I personally feel like we've said enough about Sapling at this point and should cut these instructions.

arjunkunna replied:
Removed.

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
-    a different size. (Use the "non-distributed" API below.)
- 2. Every machine in a distributed job executes an independent FFT of
-    the same size. (Use the "distributed" API below.)
+There are 4 possible modes:
```
```suggestion
Regent-FFT supports four distinct axes of usage:
```

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
+1. GPU vs. CPU
+2. Complex-to-Complex vs. Real-to-Complex
+3. Float vs. Double
+4. Batched mode (supported for all of the above)
```
```suggestion
4. Single vs. Batched
```

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
+There are several sample code snippets in the `fft_test.rg` file for reference
+as well.
+
+#### 1. Import `fft.rg`
```
```suggestion
#### 1. Import `fft.rg` and Create the FFT Interface

To generate a specific instance of the library, use `fft.generate_fft_interface(...)`.
```

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
-in. Currently only 1 to 3 dimensions and `complex64` are
-supported. For example, to initialize with 1 dimension:
+First, an FFT interface has to be generated depending on the type of transform
+you hope to do. Then, we:
```
```suggestion
you want to do. Then, we:
```

I hope we can do better than let users "hope" to use this. 😃

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
+
+#### 1. Import `fft.rg`
+
+- The first argument is the dimension: `int1d`, `int2d`, or `int3d`.
```
```suggestion
- The first argument is the index type of the input and output region: `int1d`, `int2d`, or `int3d`. This also tells the FFT interface what dimensionality to expect.
```

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
+#### 2. Make Plan
 
-Like many FFT libraries, Regent FFT requires the use of *plans*. Plans
-are specific to the sizes of the input and output regions being used,
-as well as the machine node the plan is initialized on. Currently *the
-enforcement of these assumptions is the responsibility of the user*.
+`make_plan` takes three arguments:
 
-The way that a plan is initialized depends on the usage mode. In
-general, plans are stored in a region which is managed by the
-user. The plan region may be a subregion and need not start at zero,
-but it must contain at least a number of elements depending on the
-mode: 1 in non-distributed mode, `N` in distributed mode where `N` is
-the number of nodes.
```
I think the original text here describes a lot of useful information that I would prefer not to lose. The new text describes the arguments but it doesn't say why things are the way they are or where things need to go. Please go back through this and find a better way of integrating the new and old so we're not losing information.

arjunkunna replied:
I added these paragraphs back in , in the 'make plan' section

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
+transform with double precision, the input array will have fieldspace `double`
+and output array will have fieldspace `complex64`. Larger fieldspaces that
+contain the appropriate types can also be passed in via field polymorphism -
+[here](https://groups.google.com/g/legionusers/c/yvQa8BE6QD0/m/_1cL_w-aAAAJ) is
```
There should be examples of this in the Regent test suite. Let's find one of those instead of a mailing list post.

arjunkunna replied:
Swapped this out for a link to the regent test suite: https://gitlab.com/StanfordLegion/legion/-/blob/master/language/tests/regent/run_pass/call_task_polymorphic1.rg

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
+The input region should be initialized with index space of the form
+`ispace(<type>, N)`, where N is the size of the array, and `<type>` is either
+int1d/int2d/int3d depending on the dimension of the transform. The fieldspace of
+the region is the type supported by the transform - e.g, in a real-to-complex
```
I don't have a strong opinion but you capitalized `real-to-complex` elsewhere in this README. Whatever you pick, let's be consistent about it.

arjunkunna replied:
Switched to Real-To-Complex everywhere

elliottslaughter commented on README.md:
````diff
@@ -1,189 +1,337 @@
 ```
+
+The plan region always takes the following form, with fieldspace `fft.plan` (see
+`fft.rg` for description of plan fieldspace):
````
I'm not sure we need users to be aware of what goes in `fft.plan`? Users are not intended to modify it directly.

arjunkunna replied:
That's fair. I removed this part.

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
-plan prior to loading the regions with data, or else must create a
-temporary region (of the same size and layout as the real one) for use
-in initialization.
+`make_plan` is a `__demand(__inline)` task. This means that if the user wants it
+to execute it in a separate task, they must wrap the task themselves.
```
```suggestion
Every task in the Regent-FFT interface provides two versions: one which is
`__demand(__inline)` and one which is just a regular task. Here, these are
called `make_plan` (for inline) and `make_plan_task` (for a task). Users
should pick the appropriate version depending on whether they want to launch a
new task or not.
```

arjunkunna replied:
Accepted change

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
-region and partition. The task `make_plan_distrib` is a
-`__demand(__inline)` task that internally performs an index launch
-over the machine to initialize `p`.
+Note that `execute_plan` is a `__demand(__inline)` task (similar to `make_plan`
+above). The task `execute_plan_task` is simply a wrapper around `execute_plan`
+for convenience, to avoid needing to define this explicitly.
```
Could remove now if you take my suggestion above.

arjunkunna replied:
Removed

elliottslaughter commented on README.md:
````diff
@@ -1,189 +1,337 @@
+When a plan is no longer needed, it can be destroyed:
+
+```lua
+fft1d.destroy_plan(p)
````
I think we should be consistent and call the same version for each usage of the API we're showing. So `make_plan`, `execute_plan` and `destroy_plan`; or `make_plan_task`, `execute_plan_task` and `destroy_plan_task`.

We should probably make a point about being consistent about what you do: because the processor that runs `make_plan` should probably be the same processor you `execute_plan` and `destroy_plan`.

arjunkunna replied:
In the readme, I now call `make_plan`, `execute_plan` and `destroy_plan` for the regular version and `make_plan_distrib`, `execute_plan_distrib` and `destroy_plan_distrib` for the distributed version. Also added that note about being consistent.

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
-To perform an FFT, execute the plan. Note that, unlike with
-initialization, this looks similar whether or not distributed mode is
-being used.
+To illustrate how to perform a batched transform, let us use the example where
+you want to perform 7 batches of a 256 x 256 transform.
```
Before we tell users how to do it, we should first tell them what a batched transform is.

arjunkunna replied:
Added the line in: "Batched transforms allow users to perform multiple transforms of the same size simultaneously, in a 'batch'."

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
 
-For example, if there is a single input `r` and output `s`:
+Since the transform is a 2D one, the user creates a interface with `itype` of
+dimension N+1: in this case, an `int3d`. The last dimension is used to store the
```
Didn't we change this so that the user supplies `N` and we internally add one to get `N+1` for the batched size?

arjunkunna replied:
I have now fixed this: the user supplies N, along with a flag indicating whether it is a batch transform or not, and we internally add a dimension if needed

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
-because the plans must be used by the host-side code to launch the FFT
-kernels. Therefore, when `execute_plan_task` is mapped onto the GPU it
-is very important to map the `p` region into zero-copy memory.
+Batched transforms are supported on both CPU and GPUs, for 1, 2 and 3
+dimensions. Both single-precision and double-precision real-to-complex and
+complex-to-complex transforms are supported.
```
This basically says we support all possible combinations. Therefore, I think we can just skip this.

arjunkunna replied:
Removed this.

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
+#### 6. Distributed Mode
 
-When a plan is no longer needed it can be distroyed.
+The API also supports a distributed mode, where every machine in a distributed
+job executes an independent FFT of the same size.
```
I'm not sure this is actually true? If you partitioned the input/output regions non-uniformly I think everything would work fine.

arjunkunna replied:
I think I took this from the 'implementation notes' section of the original readme. I removed the 'of the same size' portion in any case, so hopefully this is accurate now.

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
 
-In the distributed API:
+Note the use of `get_num_nodes` to determine the size of the `p` region and
+partition. The task `make_plan_distrib` is a `__demand(__inline)` task that
+internally performs an index launch over the machine to initialize `p`.
```
```suggestion
partition. The task `make_plan_distrib` is a `__demand(__inline)` task that
internally performs an index launch over the machine to initialize `p`
(i.e., it will launch one `make_plan_task` per subregion of the inputs).
```

elliottslaughter commented on README.md:
````diff
@@ -1,189 +1,337 @@
+```lua
+__demand(__index_launch)
+for i in r_part.colors do
+  fft1d.execute_plan_task(r_part[i], s_part[i], p)
````
```suggestion
  fft1d.execute_plan_task(r_part[i], s_part[i], p_part[i])
```

Very important!!! Otherwise you will get memory corruption.

elliottslaughter commented on README.md:
````diff
@@ -1,189 +1,337 @@
+end
 ```
+
+Note: `execute_plan` is a `__demand(__inline)` task (as described above). The
+task `execute_plan_task` is simply a wrapper around execute_plan for
+convenience, to avoid needing to define this explicitly.
+
+> [!IMPORTANT]
+>
+> Because execute_plan is a `__demand(__inline)` task, it will never execute on
+> the GPU (unless the parent task is running on the GPU). Therefore, in most
+> cases it is necessary to use `execute_plan_task` if one wants to use the GPU.
````
I think we've said these before? Can remove.

arjunkunna replied:
Removed .

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
+> the GPU (unless the parent task is running on the GPU). Therefore, in most
+> cases it is necessary to use `execute_plan_task` if one wants to use the GPU.
+
+> [!IMPORTANT]
+>
+> While `execute_plan_task` may be executed on the GPU, the contents of the `p`
+> region must still be available on the CPU, because the plans must be used by
+> the host-side code to launch the FFT kernels. Therefore, when
+> `execute_plan_task` is mapped onto the GPU, it is very important to map the
+> `p` region into zero-copy memory.
```
Feels like this goes elsewhere?

arjunkunna replied:
![image](https://github.com/StanfordLegion/regent-fft/assets/13012603/4fd7d58f-8e86-44a6-9964-7e375ed5e465)
I moved this up to the execute plan section.

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
-destroy the plans on each node.
+> [!NOTE]
+>
+> This is a `__demand(__inline)` task, and `destroy_plan_distrib` will
+> internally perform an index launch to destroy the plans on each node.
```
```suggestion
> As with `make_plan_distrib`, `destroy_plan_distrib` will
> internally perform an index launch to destroy the plans on each node.
```

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
-the node where the plan was originally created.
+> [!IMPORTANT]
+>
+> like `make_plan_distrib`, the index launch issued by `destroy_plan_distrib`
```
```suggestion
> Like `make_plan_distrib`, the index launch issued by `destroy_plan_distrib`
```

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
+> must be mapped so that each point task runs on the node where the plan was
+> originally created.
 
-## Caveats
```
These caveats still apply (mostly). We support GPUs now but I think everything else I wrote here originally is still applicable.

arjunkunna replied:
Added these caveats back (except for the GPU one)

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
+- Elliott Slaughter (<slaughter@cs.stanford.edu>)
+- Arjun Kunna (<arjunkunna@gmail.com>)
 
-  * Missing features:
-      * strides
-      * batches
-      * complex32
-      * backwards
-      * split
-      * real
```
Let's update the feature list instead of removing it.

arjunkunna replied:
Updated to remove batches, complex32, and real - kept the rest

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
 
-  * Multi-GPU is probably not supported right now because cuFFT plans
-    are (probably) tied to a context for a specific GPU.
+- For information on Regent, please refer to the [Regent
+  website](https://regent-lang.org/) or the [Legion
+  website](https://legion.stanford.edu/starting/).
```
We've said this multiple times so do we still need this here?

arjunkunna replied:
Removed.

elliottslaughter commented on README.md:
```diff
@@ -1,189 +1,337 @@
 
-  * FFTW's planner is *not thread-safe*, but this isn't currently
-    protected by a lock in the Regent wrapper.
+Next items in the pipeline include distributed transforms across multiple nodes.
+Please submit an issue if there are specific features that may be helpful.
```
I'd just go back to the to-do list we had before and not commit roadmap plans into a README.

arjunkunna replied:
I assume you mean the 'missing features' list above. I added that back in, and thus removed these lines

elliottslaughter commented on install.py:
```diff
@@ -52,6 +52,8 @@ def driver(thread_count):
     os.mkdir(fftw_build)
     subprocess.check_call([os.path.join(fftw_src, 'configure'), '--enable-openmp', '--disable-mpi', '--enable-shared', '--with-pic', '--prefix=%s' % fftw_install], cwd=fftw_build)
     subprocess.check_call(['make', 'install', '-j%s' % thread_count], cwd=fftw_build)
+    subprocess.check_call([os.path.join(fftw_src, 'configure'), '--enable-openmp', '--disable-mpi', '--enable-shared', '--with-pic', '--enable-single', '--prefix=%s' % fftw_install], cwd=fftw_build)
+    subprocess.check_call(['make', 'install', '-j%s' % thread_count], cwd=fftw_build)
```
I would be more comfortable if we create an entirely separate build directory for the single precision build. Who knows what state might creep over if we reuse the old one.

arjunkunna replied:
Added separate build directory: https://github.com/arjunkunna/regent-fft/commit/52fb1edb4b89819658b531ee88847918bef62fc8

elliottslaughter commented on src/fft.rg:
```diff
@@ -13,74 +13,135 @@
 
 --[[--
-Regent FFT library.
+Regent FFT library
```
I think this is a doc comment format (thus the `--[[--` weird start to this comment) so the trailing `.` is not a mistake here.

arjunkunna replied:
Added the period back.

elliottslaughter commented on src/fft.rg:
```diff
@@ -13,74 +13,135 @@
+
+  terra cufft_assert(result : cufft_c.cufftResult)
+    var status = "UNKNOWN"
+    if result == 0 then
```
Can we double check if this can be automated? I don't like hard-coding the enum values.

E.g., we might be able to do `cufft_c.CUFFT_SUCCESS`. If that works, then we just need a list of the codes and we can automate the rest of this.

arjunkunna replied:
Instead of hardcoding, switched to using tags like `cufft_c.CUFFT_SUCCESS`-- I was trying to find a dictionary like structure but this seemed like the best option https://github.com/arjunkunna/regent-fft/commit/7308dd39ae097741e3c4c7d05dbe1d7b69de6a8f

arjunkunna replied:
Resolved as discussed over email.

elliottslaughter commented on src/fft.rg:
```diff
@@ -13,74 +13,135 @@
 end
 
--- Hack: get defines from fftw3.h
+-- Define constants: get defines from fftw3.h
```
I mean, it is a hack. 😃 

The reason why the "hack" comment is here is because we cannot fetch them automatically (like I asked you to check for cuFFT). We are not "defining" the comments: that's what FFTW does. We're replicating constants that *FFTW defined*. Therefore, while it's a minor point, I'd put the "hack" comment back to reflect what we're really doing here.

arjunkunna replied:
Added this back

elliottslaughter commented on src/fft.rg:
```diff
@@ -13,74 +13,135 @@
+-- Define constants: get defines from fftw3.h
 fftw_c.FFTW_FORWARD = -1
 fftw_c.FFTW_BACKWARD = 1
-
 fftw_c.FFTW_MEASURE = 0
-fftw_c.FFTW_DESTROY_INPUT = (2 ^ 0)
-fftw_c.FFTW_UNALIGNED = (2 ^ 1)
-fftw_c.FFTW_CONSERVE_MEMORY = (2 ^ 2)
-fftw_c.FFTW_EXHAUSTIVE = (2 ^ 3) -- NO_EXHAUSTIVE is default
-fftw_c.FFTW_PRESERVE_INPUT = (2 ^ 4) -- cancels FFTW_DESTROY_INPUT
-fftw_c.FFTW_PATIENT = (2 ^ 5) -- IMPATIENT is default
 fftw_c.FFTW_ESTIMATE = (2 ^ 6)
-fftw_c.FFTW_WISDOM_ONLY = (2 ^ 21)
```
Why are these changes here? We may not use all the constants, but it seems reasonable to define everything that's available so that we don't find we need to go fish something out later.

arjunkunna replied:
Added these back

elliottslaughter commented on src/fft.rg:
```diff
@@ -13,74 +13,135 @@
 local fft = {}
 
-function fft.generate_fft_interface(itype, dtype)
+--- Create the FFT interface.
```
```suggestion
--- Create an FFT interface.
```

Since you can do it multiple times.

elliottslaughter commented on src/fft.rg:
```diff
@@ -13,74 +13,135 @@
 
-function fft.generate_fft_interface(itype, dtype)
+--- Create the FFT interface.
+-- @param itype Index type of transform (int2d/int2d/int3d/int4d).
```
```suggestion
-- @param itype Index type of transform (int1d/int2d/int3d/int4d).
```

Also I thought this was limited to 3D?

arjunkunna replied:
I indicated int4d here since for a 3D batch transform, the user has to pass in a 4D region. But this will change if we switch to the user passing in a int3d for a 3D batch transform

arjunkunna replied:
Limited to 3D now that we are only taking N and not N+1 in batched transforms

elliottslaughter commented on src/fft.rg:
```diff
@@ -131,7 +200,7 @@ function fft.generate_fft_interface(itype, dtype)
     var n = __future(int64, f)
 
-    -- FIXME (Elliott): I thought Regent was supposed to copy on
+    -- (Elliott): I thought Regent was supposed to copy on 
```
Instead of removing the FIXME let's check if this is actually still true or not. If it's still true, then this would still be a FIXME. 😃

arjunkunna replied:
Yes, this still segfaults if I uncomment that line, so added back the 'fixme'

elliottslaughter commented on src/fft.rg:
```diff
@@ -149,17 +218,17 @@ function fft.generate_fft_interface(itype, dtype)
   end
 
-  __demand(__inline)
+   __demand(__inline)
```
```suggestion
  __demand(__inline)
```

elliottslaughter commented on src/fft.rg:
```diff
@@ -149,17 +218,17 @@ function fft.generate_fft_interface(itype, dtype)
-    -- Hack: 3Bneed to use raw access to circument CUDA checker here.
-    var pr = __physical(plan)[0]
+    -- Hack: Need to use raw access to circument CUDA checker here.
+    var pr = __physical(plan)[0] -- Returns the first physical region
```
```suggestion
    var pr = __physical(plan)[0]
```

Don't transliterate the code.

elliottslaughter commented on src/fft.rg:
```diff
@@ -13,74 +13,135 @@
+  assert(dtype_in == float or dtype_in == double or dtype_in == complex32 or dtype_in == complex64, "input type must be float/double/complex32/complex64")
+  assert(dtype_out == complex32 or dtype_out == complex64, "output type must be complex32/complex64")
+
+   -- Single-precision transforms
+  local float_to_complex32_transform = (dtype_in == float and dtype_out == complex32)
+  local complex32_to_complex32_transform = (dtype_in == complex32 and dtype_out == complex32)
+
+   -- Double-precision transforms
+  local double_to_complex64_transform = (dtype_in == double and dtype_out == complex64)
+  local complex64_to_complex64_transform = (dtype_in == complex64 and dtype_out == complex64)
```
This is not the way I would want to do this. It appears that you are making variables here for the different types, and then doing *dynamic* branches in the code. That means that every interface will link to every FFT function as we do not differentiate them at runtime. It also means we always end up linking both double and single precision FFTW libraries even if we don't use both.

To me it would seem simpler to do:

```lua
local cufft_transform_type
if dtype_in == float and dtype_out == complex32 then
  cufft_transform_type = cufft_c.CUFFT_R2C
elseif ... then
  ...
else
  assert(false, "Unexpected type combination " .. tostring(dtype_in) .. " and " .. tostring(dtype_out))
end
```

And then later where you have the if statement you just:

```lua
cufft_assert(cufft_c.cufftPlanMany(&p.cufft_p, dim, &n[0], [&int](0), 0, 0, [&int](0), 0, 0, cufft_transform_type, 1))
```

No `if` statement because you branched on the FFT type earlier.

You can do the same for FFTW even though it uses different functions:

```lua
local fftw_create_plan, fftw_complex
if dtype_in == float and dtype_out == complex32 then
  fftw_create_plan = fftw_c.fftwf_plan_dft_r2c
  fftw_complex = fftw_c.fftwf_complex
elseif ... then
  ...
else
  assert(false, "Unexpected type combination " .. tostring(dtype_in) .. " and " .. tostring(dtype_out))
end
```

And then when you use it:

```lua
fftw_create_plan(dim, &n[0], [&float](input_base), [&fftw_complex](output_base), fftw_c.FFTW_ESTIMATE)
```

You can meta-program the type of the field `p` so that you don't need multiple plan fields either.

arjunkunna replied:
Have implemented this - see commit https://github.com/arjunkunna/regent-fft/commit/cf2dd5c77c68227b8083c38d5665eb067cf78c55

elliottslaughter commented on src/fft.rg:
```diff
@@ -13,74 +13,135 @@
     local terra get_base(rect : rect_t,
-                 physical : c.legion_physical_region_t,
-                 field : c.legion_field_id_t)
+                physical : c.legion_physical_region_t,
+                field : c.legion_field_id_t)
```
Hard to see in the PR interface here, but I would line this up with the start of the `rect` parameter. (If you want to fix the spacing. Or else I'd remove the diff entirely and let it go back to the original, which is still wrong, but at least doesn't cause a diff for something you're not fixing anyway.)

arjunkunna replied:
Lined this up.

elliottslaughter commented on src/fft.rg:
```diff
@@ -94,26 +155,34 @@ function fft.generate_fft_interface(itype, dtype)
       end
       regentlib.assert(offsets[0].offset == terralib.sizeof(t), "stride does not match expected value")
-
+      
```
```suggestion

```

elliottslaughter commented on src/fft.rg:
```diff
@@ -172,105 +241,279 @@ function fft.generate_fft_interface(itype, dtype)
-        regentlib.assert(ok == cufft_c.CUFFT_SUCCESS, "cufftPlanMany failed")
-        return cufft_p
-      else
+      if c.legion_processor_kind(proc) ~= c.TOC_PROC then
         regentlib.assert(false, "make_plan_gpu must be executed on a GPU processor")
```
We could wrap this into the assert.

arjunkunna replied:
Wrapped.

elliottslaughter commented on src/fft.rg:
```diff
@@ -172,105 +241,279 @@ function fft.generate_fft_interface(itype, dtype)
+      var lo = input.ispace.bounds.lo:to_point()
+      var hi = input.ispace.bounds.hi:to_point()
+      var n : int[dim] -- n is an array of size dim with the size of each dimension in the entries
+      ;[data.range(dim):map(function(i) return rquote n[i] = hi.x[i] - lo.x[i] + 1 end end)]
```
If this is truly not needed, let's delete it?

arjunkunna replied:
I don't think the semicolon was intended to comment out this line. This line is definitely needed. Removing the semicolon

arjunkunna replied:
Added semi colon back. Resolved as discussed over email. This line is needed.

elliottslaughter commented on src/fft.rg:
```diff
@@ -172,105 +241,279 @@ function fft.generate_fft_interface(itype, dtype)
+      var output_base = get_base_out(rect_out_t(output.ispace.bounds), __physical(output)[0], __fields(output)[0]).base
+      var lo = input.ispace.bounds.lo:to_point()
+      var hi = input.ispace.bounds.hi:to_point()
+      var n : int[dim] -- n is an array of size dim with the size of each dimension in the entries
```
Most of this comment is transliterating the code. But what is `n`, really? It's the size of the input region? Something else?

arjunkunna replied:
n is meant to be as follows: n[In] – Array of size rank, describing the size of each dimension, n[0] being the size of the outermost and n[rank-1] innermost (contiguous) dimension of a transform.

(from https://docs.nvidia.com/cuda/cufft/#cufftplanmany)

elliottslaughter commented on src/fft.rg:
```diff
@@ -172,105 +241,279 @@ function fft.generate_fft_interface(itype, dtype)
-  __demand(__inline)
-  task iface.make_plan(input : region(ispace(itype), dtype),
-                       output : region(ispace(itype), dtype),
+  --- Creates the FFT plan. This is the __inline version of the task should the user wish to use that. Entry point into GPU functionality as well.
```
```suggestion
  --- Creates the FFT plan. This __inline version will execute in the caller's context. Supports both CPU and GPU.
```

elliottslaughter commented on src/fft.rg:
```diff
@@ -172,105 +241,279 @@ function fft.generate_fft_interface(itype, dtype)
+  -- @param input Input region.
+  -- @param output Output region.
+  -- @param plan Plan region.
+  -- @note Calls `make_plan_gpu` if necessary.
```
```suggestion
```

`make_plan_gpu` is not part of the public interface and we should never mention it in the documentation or API comments.

elliottslaughter commented on src/fft.rg:
```diff
@@ -172,105 +241,279 @@ function fft.generate_fft_interface(itype, dtype)
-    var address_space = c.legion_processor_address_space(
-      get_executing_processor(__runtime()))
+    var n : int[dim]
+    ;[data.range(dim):map(function(i) return rquote n[i] = hi.x[i] - lo.x[i] + 1 end end)]
```
Same about stray comments. Either we need this or we don't, and if the latter, just get rid of it.

arjunkunna replied:
Same as above. removing the semicolon but keeping this line

elliottslaughter commented on src/fft.rg:
```diff
@@ -172,105 +241,279 @@ function fft.generate_fft_interface(itype, dtype)
 
-  task iface.make_plan_task(input : region(ispace(itype), dtype),
-                            output : region(ispace(itype), dtype),
+  --- Make plan: task version. As make_plan is a __demand(__inline) task, we provide make_plan_task as a wrapper for convenience, should the user wish to use a new task and not an inlined one.
```
```suggestion
  --- Creates the plan. This version will launch a task.
```

What I'm doing in the suggestion is make the documentation consistent with other versions and shortening to focus on what users need to know, without extraneous details.

elliottslaughter commented on src/fft.rg:
```diff
@@ -172,105 +241,279 @@ function fft.generate_fft_interface(itype, dtype)
     iface.make_plan(input, output, plan)
   end
 
+  --- Make plan (distributed version). This is intended to be called from inside the user's main and avoids the need for the user to directly (say) index launch make_plan_task repeatedly.
```
```suggestion
  --- Create the plan. This version launches `make_plan_task` for each subregion in the provided partitions.
```

elliottslaughter commented on test.sh:
```diff
@@ -0,0 +1,30 @@
+
+export CUDA="/usr/local/cuda"
+export CUDA_PATH="/usr/local/cuda"
+export PATH="$CUDA_PATH/bin:$PATH"
+export LD_LIBRARY_PATH="$CUDA_PATH/lib:$CUDA_PATH/lib64:$LD_LIBRARY_PATH"
```
Why first?

arjunkunna replied:
Swapped this order as discussed over email.

elliottslaughter commented on test/fft_test.rg:
```diff
@@ -17,92 +17,415 @@ import "regent"
+  local t = regentlib.newsymbol(title, "t")
+  local i = regentlib.newsymbol(input, "i")
+  local task print_region_task([t], [i])
+  where reads (i) do
```
```suggestion
  where reads(i) do
```

elliottslaughter commented on test/fft_test.rg:
```diff
@@ -17,92 +17,415 @@ import "regent"
+  local o = regentlib.newsymbol(output, "o")
+  local e = regentlib.newsymbol(expected, "e")
+  local task compare_region_task([o], [e])
+  where reads (o, e) do
```
```suggestion
  where reads(o, e) do
```

elliottslaughter commented on test/fft_test.rg:
```diff
@@ -17,92 +17,415 @@ import "regent"
+  local e = regentlib.newsymbol(expected, "e")
+  local task compare_region_task([o], [e])
+  where reads (o, e) do
+    var status = "PASSED"
```
Why use a string here? It seems more reasonable to calculate and return a boolean and manage how we print the result to the user on the caller's side. Plus, I'm 99.9% sure this will crash on multiple nodes.

arjunkunna replied:
Added a boolean and asserts: https://github.com/arjunkunna/regent-fft/commit/686a130f6f12c8154a24511b58248d6a22506470

elliottslaughter commented on test/fft_test.rg:
```diff
@@ -17,92 +17,415 @@ import "regent"
+  print_region_1d_complex32("Expected", e)
+
+  var status = compare_regions_1d_complex32(s, e)
+  format.println("<< test_1d_complex32_to_complex32_transform [{}]", status)
```
Like I said above I think we should make the `status` here a boolean. Also, this test as written will not actually fail the process if it does not pass. We need an assert so that a failing test will actually fail the process. Otherwise we'll never see the failures. It's still ok to print if you want, we just need the assert to see the failure.

arjunkunna replied:
Added a boolean and asserts: https://github.com/arjunkunna/regent-fft/commit/686a130f6f12c8154a24511b58248d6a22506470

elliottslaughter commented on test/test_mapper.cc:
```diff
@@ -24,100 +26,46 @@ using namespace Legion::Mapping;
 
-class FFTTestMapper : public DefaultMapper
-{
+// extend Default mapper
```
Don't transliterate the code.

arjunkunna replied:
Removed these comments - they were also absent in the original repo.

elliottslaughter reviewed (commented):
Thanks again for working on the last round of edits. We're getting there, but there are a new batch of comments here to work through before this will be ready.

arjunkunna committed:
Update README.md

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

arjunkunna committed:
Update README.md

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

arjunkunna committed:
Update test/fft_test.rg

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

arjunkunna committed:
Update test/fft_test.rg

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

arjunkunna committed:
Update README.md

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

arjunkunna committed:
Update README.md

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

arjunkunna committed:
Update README.md

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

arjunkunna committed:
Update README.md

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

arjunkunna committed:
Update README.md

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

arjunkunna committed:
Update README.md

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

arjunkunna committed:
Update README.md

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

arjunkunna committed:
Update README.md

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

arjunkunna committed:
Update README.md

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

arjunkunna committed:
Update README.md

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

arjunkunna committed:
Update README.md

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

arjunkunna committed:
Update src/fft.rg

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

arjunkunna committed:
Update src/fft.rg

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

arjunkunna committed:
Update README.md

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

arjunkunna committed:
Update README.md

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

arjunkunna committed:
Update src/fft.rg

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

arjunkunna committed:
Update src/fft.rg

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

arjunkunna committed:
Update src/fft.rg

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

arjunkunna committed:
Update README.md

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

arjunkunna committed:
Update README.md

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

arjunkunna committed:
Update src/fft.rg

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

arjunkunna committed:
Update src/fft.rg

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

arjunkunna committed:
Update src/fft.rg

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

arjunkunna committed:
Update README.md

arjunkunna committed:
Update README.md

arjunkunna committed:
Update README.md

arjunkunna committed:
Update README.md

arjunkunna committed:
Update README.md

arjunkunna committed:
Update README.md

arjunkunna committed:
Update README.md

arjunkunna committed:
Update README.md

arjunkunna committed:
Update README.md

Commit:
resolving some comments from elliott

arjunkunna committed:
Update README.md

arjunkunna committed:
Update README.md

arjunkunna committed:
Update README.md

Commit:
addressing comments

arjunkunna committed:
clang-format: git diff should fail for format changes (added --exit-code)

arjunkunna committed:
Replace cufftResult hardcoded enum values with their respective identifiers

arjunkunna committed:
Add condition in regentlib.assert (combine asserts)

arjunkunna committed:
Add assertions to verify tests

arjunkunna committed:
Use separate build directory for single precision FFTW

Commit:
edited assert language

arjunkunna committed:
Update README.md

arjunkunna committed:
Update README.md

Commit:
distrib version tweaks

Commit:
change batch to use plan of size n

Commit:
trim trailing whitspace

arjunkunna committed:
Update README.md

arjunkunna committed:
Update README.md

arjunkunna committed:
Update README.md

arjunkunna committed:
Update README.md

arjunkunna committed:
Update README.md

arjunkunna committed:
Update README.md

arjunkunna committed:
Update README.md

Commit:
additional comment addressing

Commit:
Merge branch 'master' of https://github.com/arjunkunna/regent-fft

arjunkunna committed:
Metaprogam CUFFT's plan APIs

arjunkunna committed:
Metaprogram CUFFT's execute APIs

arjunkunna committed:
Metaprogram FFTW's plan and execute APIs

arjunkunna committed:
Load FFTW single/double libs conditionally

elliottslaughter commented on test/fft_test.rg:
```diff
@@ -17,92 +17,420 @@ import "regent"
+  return print_region_task
+end
+
+local print_region_1d_float = make_print_region_task(rawstring, region(ispace(int1d), float))
+local print_region_1d_double = make_print_region_task(rawstring, region(ispace(int1d), double))
+local print_region_1d_complex32 = make_print_region_task(rawstring, region(ispace(int1d), complex32))
+local print_region_1d_complex64 = make_print_region_task(rawstring, region(ispace(int1d), complex64))
+local print_region_2d_complex64 = make_print_region_task(rawstring, region(ispace(int2d), complex64))
+local print_region_3d_double = make_print_region_task(rawstring, region(ispace(int3d), double))
+local print_region_3d_complex64 = make_print_region_task(rawstring, region(ispace(int3d), complex64))
+local print_region_4d_complex64 = make_print_region_task(rawstring, region(ispace(int4d), complex64))
```
Can we change `rawstring` to `regentlib.string`? Technically `rawstring` is not safe to pass to a task because the pointer may not be valid across nodes.

arjunkunna replied:
Edited here: https://github.com/arjunkunna/regent-fft/commit/103303a67e88d24b221b1de468e6d7bf2cc806d5

elliottslaughter commented on src/fft.rg:
```diff
@@ -48,39 +91,168 @@ fftw_c.FFTW_WISDOM_ONLY = (2 ^ 21)
+    elseif itype == int2d then
+      itype = int3d
+    elseif itype == int3d then
+      itype = int4d
```
```
else
  assert(false)
```

To make sure we don't forget any cases here.

arjunkunna replied:
Edited here: https://github.com/arjunkunna/regent-fft/commit/103303a67e88d24b221b1de468e6d7bf2cc806d5

elliottslaughter commented on src/fft.rg:
```diff
@@ -48,39 +91,168 @@ fftw_c.FFTW_WISDOM_ONLY = (2 ^ 21)
+  elseif double_precision then
+    regentlib.linklibrary("libfftw3.so")
+    fftw_plan_handle_type = fftw_c.fftw_plan
+    fftw_destroy_plan_function = fftw_c.fftw_destroy_plan
```
`else assert(false)`

Being pedantic, but might as well be safe here.

arjunkunna replied:
Edited here: https://github.com/arjunkunna/regent-fft/commit/103303a67e88d24b221b1de468e6d7bf2cc806d5

elliottslaughter commented on src/fft.rg:
```diff
@@ -172,105 +351,235 @@ function fft.generate_fft_interface(itype, dtype)
+      elseif dim == 3 then
+        i_dist = offset_3 / offset_1
+      elseif dim == 4 then
+        i_dist = offset_in[3].offset / offset_1
```
`else regentlib.assert(...)`?

arjunkunna replied:
Edited here: https://github.com/arjunkunna/regent-fft/commit/103303a67e88d24b221b1de468e6d7bf2cc806d5

elliottslaughter commented on src/fft.rg:
```diff
@@ -172,105 +351,235 @@ function fft.generate_fft_interface(itype, dtype)
+    elseif dim == 3 then
+      i_dist = offset_3/offset_1
+    elseif dim == 4 then
+      i_dist = offset_in[3].offset/offset_1
```
`else regentlib.assert(...)`?

arjunkunna replied:
Edited here: https://github.com/arjunkunna/regent-fft/commit/103303a67e88d24b221b1de468e6d7bf2cc806d5

elliottslaughter commented on src/fft.rg:
```diff
@@ -279,48 +588,148 @@ function fft.generate_fft_interface(itype, dtype)
+                          output : region(ispace(itype), dtype_out),
+                          plan : region(ispace(int1d), iface.plan),
+                          address_space : c.legion_address_space_t)
+    where reads writes (input, output, plan) do
```
```suggestion
    where reads writes(input, output, plan) do
```

arjunkunna replied:
Accepted your change

elliottslaughter commented on src/fft.rg:
```diff
@@ -279,48 +588,148 @@ function fft.generate_fft_interface(itype, dtype)
+                          output : region(ispace(itype), dtype_out),
                           plan : region(ispace(int1d), iface.plan))
-  where reads(input, plan), writes(output) do
+  where reads writes (input, output, plan) do
```
Please double check that these privileges are correct. Are we writing `input`? `plan`?

Also, cosmetic but normally no space after the privilege and before `(`. So `reads(...)` or `reads writes(...)` but not `reads (...)` or `reads writes (...)`.

arjunkunna replied:
Changed here: https://github.com/arjunkunna/regent-fft/commit/ba08bf54876149a276a03d839b6e85f4f7e5a7e8 (and add on here https://github.com/arjunkunna/regent-fft/commit/06de4bd456bf0f2f5dfa11d7d807a8d667613d23)

elliottslaughter commented on src/fft.rg:
```diff
@@ -279,48 +588,148 @@ function fft.generate_fft_interface(itype, dtype)
+                               output : region(ispace(itype), dtype_out),
                                plan : region(ispace(int1d), iface.plan))
-  where reads(input, plan), writes(output) do
+  where reads writes (input, output, plan) do
```
Same note to check the privileges on all of these execute task versions.

arjunkunna replied:
Changed here: https://github.com/arjunkunna/regent-fft/commit/ba08bf54876149a276a03d839b6e85f4f7e5a7e8 (and add on here https://github.com/arjunkunna/regent-fft/commit/06de4bd456bf0f2f5dfa11d7d807a8d667613d23)

elliottslaughter commented on CHANGES.md:
```diff
@@ -0,0 +1,7 @@
+# Version History
+
+- 1.0: Initial Release (Apr 2024)
```
```suggestion
- 1.0: Initial Release (Jul 2024)
```

arjunkunna replied:
Accepted your change

elliottslaughter reviewed (approved):
This looks good. There are only a few minor things I wanted to point out:

 * A few places you could `else assert(...)` to make sure we don't miss any cases.
 * A few places I wanted you to double check the privileges to make sure we have them right.

Thanks again for working on this and for putting up with so many comments!

arjunkunna committed:
Update CHANGES.md

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

arjunkunna committed:
Update src/fft.rg

Co-authored-by: Elliott Slaughter <elliottslaughter@gmail.com>

Commit:
Additional asserts and changing rawstring to regentlib.string

Commit:
changing privileges for execute tasks

Commit:
removed extra words

arjunkunna committed:
Update assert messages
